lib/events/athena/consumer.go

```diff
@@ -0,0 +1,457 @@
+	// TODO(tobiaszheller): delete messages from queue in next PR.
+}
+
+// sqsMessagesCollector is responsible for collecting messages from SQS and
+// writing to on channel.
+type sqsMessagesCollector struct {
+	logger        log.FieldLogger
+	config        sqsCollectConfig
+	errHandlingFn func(ctx context.Context, errC chan error)
+	eventsChan    chan eventAndAckID
+}
+
+// newSqsMessagesCollector returns message collector.
+// Collector sends collected messages from SQS on events channel.
+func newSqsMessagesCollector(cfg sqsCollectConfig, log log.FieldLogger, errHandlingFn func(ctx context.Context, errC chan error)) *sqsMessagesCollector {
+	return &sqsMessagesCollector{
+		logger:        log,
+		config:        cfg,
+		errHandlingFn: errHandlingFn,
+		eventsChan:    make(chan eventAndAckID, cfg.batchMaxItems),
+	}
+}
+
+// getEventsChan returns channel which can be used to read messages from SQS.
+// When collector finishes, channel will be closed.
+func (s *sqsMessagesCollector) getEventsChan() <-chan eventAndAckID {
+	return s.eventsChan
+}
+
+// fromSQS receives messages from SQS and sends it on eventsC channel.
+// It runs until context is canceled (via timeout) or when maxItems is reached.
+// MaxItems is soft limit and can happen that it will return more items then MaxItems.
+func (s *sqsMessagesCollector) fromSQS(ctx context.Context) {
+	// Errors should be immediately process by error handling loop, so 10 size
+	// should be enough to not cause blocking.
+	errorsC := make(chan error, 10)
+	defer close(errorsC)
+
+	// errhandle loop for receiving single event errors.
+	go func() {
+		s.errHandlingFn(ctx, errorsC)
+	}()
+	eventsC := s.eventsChan
+
+	count := 0
+	countMu := sync.Mutex{}
+
+	// wokerCtx is mechanism to stop other workers when maxItems is reached.
+	wokerCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	wg := sync.WaitGroup{}
+	wg.Add(maxNumberOfWorkers)
+	for i := 0; i < maxNumberOfWorkers; i++ {
+		go func(i int) {
+			defer wg.Done()
+			for {
+				if wokerCtx.Err() != nil {
+					return
+				}
+
+				// If there is not enough time to process receiveMessage call
+				// we can return immediately. It's added because if
+				// receiveMessages is canceled message is marked as not
+				// processed after VisibilitTimeout (equal to BatchInterval).
+				if deadline, ok := wokerCtx.Deadline(); ok && time.Until(deadline) <= s.config.waitOnReceiveDuration {
+					return
+				}
+				events, err := s.receiveMessages(wokerCtx)
+				if err != nil {
+					// TODO(tobiaszheller): maybe we need to check for other errors than cancel as well
+					// from aws SDK.
+					if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
+						return
+					}
+					errorsC <- trace.Wrap(err)
+				}
+				for _, e := range events {
+					if e.err != nil {
+						errorsC <- e.err
+						continue
+					}
+					eventsC <- eventAndAckID{
+						event:         e.data,
+						receiptHandle: e.receiptHandle,
+					}
+				}
```

Comment:
In `receiveMessages` you're allocating a slice then looping over it here and writing to `errorsC` and `eventsC`.

What about passing `errorsC` and `eventsC` into `receiveMessages` and writing to those channels directly? You would save the slice allocation and reduce your lines of code.

Why not pass  into

Reply:
solved in https://github.com/gravitational/teleport/pull/24038/commits/129924c4ad13c55cc2cf9409962b92225c9f65f8